### PR TITLE
URLSession additions to allow secure communications in SwiftyRequest 

### PR DESCRIPTION
**Pull request by ShihabMehboob**
Amended the code in the RestRequest.swift file to check for whether the URL passed in is a secure connection (HTTPS) or not (HTTP), after which it makes use of the URLSession delegate and its corresponding authentication functions if it is. A test case has been added too to check the connection with a HTTPS URL. Additionally, the code was attempted in an iOS app (iOSSampleKituraKit) with amendments made to the contained SwiftyRequest class, and the Application.swift file in the server folder to allow using the self-signed certificate (this is only client-side and didn't have any changes in the RestRequest.swift file).

### CODE REVIEW

**Comment by rolivieri** [comment]
@sandmman  As part of this review, adding here a link to a discussion @ShihabMehboob  and I are having in regards to this PR: https://github.ibm.com/arf/planning-swift-solutions/issues/524#issuecomment-4186988.

**Comment by ShihabMehboob** [comment]
Updated the PR with an optional Bool with a default value for the RestRequest initializer. The end user could get this value from their own config file or hardcode it if they wish to do so, allowing them to make use of the HTTPS load code if they have a self-signed certificate.

File: Sources/SwiftyRequest/RestRequest.swift [hunk 1]
```diff
@@ -44,15 +56,28 @@ public class RestRequest {
...
         }
     }
-
+    
+    /// URLSession delegate for secure loads
+    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        let method = challenge.protectionSpace.authenticationMethod
+        let host = challenge.protectionSpace.host
+        switch (method, host) {
+        case (NSURLAuthenticationMethodServerTrust, "https://localhost:8080"):
```
Comment (sandmman) [inline Sources/SwiftyRequest/RestRequest.swift]: Unless I'm mistaken, I think we need to find a way to enable users to declare the urls that should be trusted rather than defaulting to a specific host and port.

Though much more complicated than we might need, Alamofire allows users to describe the server trust policies for specific urls. See an example [here](https://stackoverflow.com/questions/32553414/alamofire-with-a-self-signed-certificate-servertrustpolicy). Perhaps we want something more along these lines?

**Comment by ShihabMehboob** [comment]
@sandmman Updated the code with the changes mentioned, will look into trusting varying ports akin Alamofire.

**Comment by rolivieri** [comment]
@ShihabMehboob Hey, can you check the build? It is failing on macOS and linux: https://travis-ci.org/IBM-Swift/SwiftyRequest/builds/316356018?utm_source=github_status&utm_medium=notification.

**Comment by rolivieri** [comment]
@ShihabMehboob I was checking the status of the latest build and it looks like we still have a failure, this time on linux: https://travis-ci.org/IBM-Swift/SwiftyRequest/builds/318069024

**Comment by ShihabMehboob** [comment]
@rolivieri `serverTrust` doesn't exist on `URLProtectionSpace` on Linux. I'm thinking I should put a check around that code block to only run for macOS. However, this'll mean that trusting certain URLs won't work on Linux builds.

File: Sources/SwiftyRequest/RestRequest.swift [hunk 2]
```diff
@@ -18,17 +18,29 @@ import Foundation
 import CircuitBreaker
 
 /// Object containing everything needed to build HTTP requests and execute them
-public class RestRequest {
-
+public class RestRequest: NSObject  {
+    
+    // Check if there exists a self-signed certificate and whether it's a secure connection
+    private var isSecure: Bool = false
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob Can we make `isSecure` a constant instead? Looking at the code, it seems we should be able to do that: `private let isSecure: Bool`... and then assign the corresponding value in the initializer.

File: Sources/SwiftyRequest/RestRequest.swift [hunk 3]
```diff
@@ -18,17 +18,29 @@ import Foundation
...
+    private var isSelfSigned: Bool = false
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob Can we make `isSelfSigned ` a constant instead? Looking at the code, it seems we should be able to do that: `private let isSelfSigned: Bool`... and then assign the corresponding value in the initializer.

File: Sources/SwiftyRequest/RestRequest.swift [hunk 4]
```diff
@@ -740,3 +756,26 @@ public enum RestError: Error, CustomStringConvertible {
...
+    
+    /// URL session function to allow trusting certain URLs
+    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        let method = challenge.protectionSpace.authenticationMethod
+        let host = challenge.protectionSpace.host
+        switch (method, host) {
+        case (NSURLAuthenticationMethodServerTrust, self.url):
+            #if MAC_OS_X_VERSION_10_6
+                let trust = challenge.protectionSpace.serverTrust
+                let credential = URLCredential(trust: trust!)
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob I am not very familiar with this API so bear with me. I am looking at like 771 and I see we are  force unwrapping the `SecTrust ` variable returned when accessing the `serverTrust` field: https://developer.apple.com/documentation/foundation/urlauthenticationchallenge/1410012-protectionspace

If in the code above, we have the possibility of `trust` being nil, then we should not force unwrap it. That would cause a runtime error and a crash. Can you double check whether nil could be returned in our case?

File: Sources/SwiftyRequest/RestRequest.swift [hunk 5]
```diff
@@ -740,3 +756,26 @@ public enum RestError: Error, CustomStringConvertible {
...
+// URL Session extension
+extension RestRequest: URLSessionDelegate {
+    
+    /// URL session function to allow trusting certain URLs
+    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        let method = challenge.protectionSpace.authenticationMethod
+        let host = challenge.protectionSpace.host
+        switch (method, host) {
+        case (NSURLAuthenticationMethodServerTrust, self.url):
+            #if MAC_OS_X_VERSION_10_6
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob A few points regarding the issue that `challenge.protectionSpace.serverTrust` is not available on Linux:

- When was `challenge.protectionSpace.serverTrust` introduced to the macOS API? Is it only available on macOS VERSION_10_6 and later? If it was recently introduced, then I am wondering what APIs were used prior to this addition in order to achieve the same goal/behavior we need.

- Is there another way (e.g. different API methods) for us to achieve the same goal and be platform neutral? It may be worth to poke around the API docs and see if there are any alternatives.

- If in the end, this is the only viable solution, I recommend we log warnings in this section of the code whenever we are running on a platform that does not have yet the necessary API implementations... but I am hoping there is a platform neutral solution.

- Either way, we should notify the foundation folks about the discrepancy in the API (macOS vs linux). I recommend you share these details... @ianpartridge or @pushkarnk can probably be the first point of contact?

**Comment by ShihabMehboob** [comment]
@rolivieri I've amended the file to include your requested changes, and started exploring the `challenge.protectionSpace.serverTrust` issue. After some research, it's evident that it doesn't not exist on Linux and therefore cannot directly be used, but I have attempted to use `SecTrust` and accompanying methods to try and get it to work on Linux. Attempting to see whether travis passes with the latest commit.

**Comment by rolivieri** [comment]
@ShihabMehboob Thanks for the update. I just verified how old macOS 10.6 is... and it is pretty old. MacOS 10.6 corresponds to Mac OS X Snow Leopard, which was released in 2009. Given that, I think we should be OK (from a macOS perspective). If someone is using our code and an older version of macOS... well, they should upgrade ASAP.

Now regarding linux, that's too bad we don't have those same capabilities yet implemented in the Foundation APIs. If you haven't yet, can you reach out to @ianpartridge and find out what is the best way to report this problem so it can be addressed by the folks who work on Foundation? 

In the meantime, yeah, at a minimum we should generate warning logs. I see you added a few print statements. However, let's make a few changes to that. I will provide details on this soon.

File: Sources/SwiftyRequest/RestRequest.swift [hunk 6]
```diff
@@ -740,3 +756,32 @@ public enum RestError: Error, CustomStringConvertible {
...
+                let trust = challenge.protectionSpace.serverTrust
+                if let trust = trust {
+                    let credential = URLCredential(trust: trust)
+                    completionHandler(.useCredential, credential)
+                } else {
+                    print("Attempting to establish a secure connection; no server trust established. Resorting to default handling.")
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob Let's replace all print statements with logging statements that use our LoggerAPI. That way we can assign a log level to each log statement. As reference, see what we did for FileKit: https://github.com/IBM-Swift/FileKit/blob/master/Package.swift#L16, https://github.com/IBM-Swift/FileKit/blob/master/Sources/FileKit/FileKit.swift#L44.

File: Sources/SwiftyRequest/RestRequest.swift [hunk 7]
```diff
@@ -740,3 +756,32 @@ public enum RestError: Error, CustomStringConvertible {
...
+    
+    /// URL session function to allow trusting certain URLs
+    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        let method = challenge.protectionSpace.authenticationMethod
+        let host = challenge.protectionSpace.host
+        switch (method, host) {
+        case (NSURLAuthenticationMethodServerTrust, self.url):
+            #if MAC_OS_X_VERSION_10_6
+                let trust = challenge.protectionSpace.serverTrust
+                if let trust = trust {
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob Could this be simplified to `if let trust = challenge.protectionSpace.serverTrust {...`?

**Comment by ShihabMehboob** [comment]
@rolivieri These changes have now been included, and replaced print statements with LoggerAPI. And yes I don't think anyone will or should be using Snow Leopard given that it's almost nine years old.

In regards to this working on Linux, talking to @ianpartridge confirmed that it does not yet exist, but work has been started on it by @pushkarnk and should be available in the next major Swift version.

File: Package.swift [hunk 8]
```diff
@@ -27,12 +27,13 @@ let package = Package(
         )
     ],
     dependencies: [
-      .package(url: "https://github.com/IBM-Swift/CircuitBreaker.git", .upToNextMajor(from: "3.0.0"))
+        .package(url: "https://github.com/IBM-Swift/CircuitBreaker.git", .upToNextMajor(from: "3.0.0")),
```
Comment (rolivieri) [inline Package.swift]: @ShihabMehboob We just released a newer version of CircuitBreaker (5.0.0). Can you update Package.swift accordingly so we can leverage that version in this PR?

File: Sources/SwiftyRequest/RestRequest.swift [hunk 9]
```diff
@@ -740,3 +757,31 @@ public enum RestError: Error, CustomStringConvertible {
...
+            #if MAC_OS_X_VERSION_10_6
+                if let trust = challenge.protectionSpace.serverTrust {
+                    let credential = URLCredential(trust: trust)
+                    completionHandler(.useCredential, credential)
+                } else {
+                    Log.debug("Attempting to establish a secure connection; no server trust established. Resorting to default handling.")
+                    completionHandler(.performDefaultHandling, nil)
+                }
+            #else
+                Log.debug("Attempting to establish a secure connection; macOS 10.6 or higher must be used to achieve this. Resorting to default handling.")
```
Comment (rolivieri) [inline Sources/SwiftyRequest/RestRequest.swift]: @ShihabMehboob Can we change the log level from debug to warning in these lines (775, 779)?

File: Sources/SwiftyRequest/RestRequest.swift [hunk 10]
```diff
@@ -577,80 +595,82 @@ public class RestRequest {
...
-    private func handleInvocation(invocation: Invocation<(Data?, HTTPURLResponse?, Error?) -> Void, Void, String>) {
+    private func handleInvocation(invocation: Invocation<(String), String>) {
         let task = session.dataTask(with: request) { (data, response, error) in
             if error != nil {
-                invocation.notifyFailure()
+                invocation.notifyFailure(error: BreakerError.fastFail)
             } else {
                 invocation.notifySuccess()
             }
-            let callback = invocation.commandArgs
```
Comment (sandmman) [inline Sources/SwiftyRequest/RestRequest.swift]: Why were these deleted? Unless there's something I'm missing, we need to pass the callback to through the invocation and then call it with the datatask result

File: Sources/SwiftyRequest/RestRequest.swift [hunk 11]
```diff
@@ -16,43 +16,56 @@
...
+        }
+        return session
+    }
+    
     // The HTTP Request
     private var request: URLRequest
-
+    
     /// `CircuitBreaker` instance for this `RestRequest`
-    public var circuitBreaker: CircuitBreaker<(Data?, HTTPURLResponse?, Error?) -> Void, Void, String>?
```
Comment (sandmman) [inline Sources/SwiftyRequest/RestRequest.swift]: The first type declaration in for the CircuitBreaker is the invocation command argument. We want to pass `(Data?, HTTPURLResponse?, Error?) -> Void`, which can then be used to pass back the URLRequest result. 

The second type declaration can be removed (`Void`). It denoted the result type of the command execution. Since all commands are now asynchronous in CircuitBreaker, we removed the need to explicitly state it.

`CircuitBreaker<(Data?, HTTPURLResponse?, Error?) -> Void, String>

File: Sources/SwiftyRequest/RestRequest.swift [hunk 12]
```diff
@@ -155,45 +168,50 @@ public class RestRequest {
...
+    
     // MARK: Response methods
-
     /// Request response method that either invokes `CircuitBreaker` or executes the HTTP request
     ///
     /// - Parameter completionHandler: Callback used on completion of operation
     public func response(completionHandler: @escaping (Data?, HTTPURLResponse?, Error?) -> Void) {
         if let breaker = circuitBreaker {
-            breaker.run(commandArgs: completionHandler, fallbackArgs: "Circuit is open")
+            breaker.run(commandArgs: "test", fallbackArgs: "Circuit is open")
```
Comment (sandmman) [inline Sources/SwiftyRequest/RestRequest.swift]: We need the completion handler. Also, as an aside, when we need string descriptions in our code, we should try to be more descriptive.

**Comment by sandmman** [comment]
@ShihabMehboob Just as an fyi, CircuitBreaker now calls the fallback method on all command failures in addition to timeout errors and when failing fast. This means we'll have to update some of the test cases that previously were dependent on the fallback being called after the max number of failures occurs (the circuit becomes open).

File: Sources/SwiftyRequest/RestRequest.swift [hunk 13]
```diff
@@ -577,80 +595,84 @@ public class RestRequest {
...
+    
     /// Method used by `CircuitBreaker` as the contextCommand
     ///
     /// - Parameter invocation: `Invocation` contains a command argument, Void return type, and a String fallback arguement
-    private func handleInvocation(invocation: Invocation<(Data?, HTTPURLResponse?, Error?) -> Void, Void, String>) {
+    private func handleInvocation(invocation: Invocation<(Data?, HTTPURLResponse?, Error?) -> Void, String>) {
         let task = session.dataTask(with: request) { (data, response, error) in
             if error != nil {
-                invocation.notifyFailure()
+                invocation.notifyFailure(error: BreakerError.fastFail)
```
Comment (sandmman) [inline Sources/SwiftyRequest/RestRequest.swift]: This shouldn't be a fastFail error. This is reserved for errors caused by the circuit being open; instead, we should  create a custom BreakerError and pass that in
```
invocation.notifyFailure(error: BreakerError(reason: error.description))
```

File: Sources/SwiftyRequest/RestRequest.swift (hunk 9 above)
Comment (sandmman) [reply Sources/SwiftyRequest/RestRequest.swift]: bump
